### PR TITLE
Add snapshot testing for pipeline files

### PR DESCRIPTION
**Pull request by bourdakos1**
This PR adds two tests:
1. Check the JSON of a pipeline that had all the nodes deleted
2. Check the JSON of a pipeline that had one node added

The bulk of this PR adds a snapshotting mechanism to make these tests easier to write and update.

## Basic usage
```js
cy.readFile('example.pipeline').matchesSnapshot();
```

The first run of this in a test will generate a snapshot of the pipeline file if it does not exist. If it exists, it will verify that the file matches the stored snapshot. If the test cannot find a snapshot in CI, the test will fail instead of generating a new snapshot.

In order to update a snapshot, delete the snapshot and re-run the integration tests for a new one to be generated.  

## Highlights of changes to pipeline json

changes going from 2.2.4 -> master (complex example):

These are some of the diffs for the snapshot of the complex pipeline example

the elyra-example-anaconda runtime didn't exist so I swapped it for the only only anaconda option:
```diff
- "runtime_image": "elyra/elyra-example-anaconda-py3:latest",
+ "runtime_image": "continuumio/anaconda3:2020.07",
```

added the new pipeline app_data:
```diff
+  "properties": {
+    "name": "simple",
+    "runtime": "Generic",
+  },
```

updated notebook description (python script description hasn't changed):
```diff
-  "description": "Notebook file",
+  "description": "Notebook",
```

When the node doesn't have any errors, it no longer has a null `invalidNodeError` property:
```diff
- "invalidNodeError": null,
```

There is a new label property:
```diff
+ "label": "",
```
> **Note:** Labels are handled differently in the new `pipeline-editor` previously we were manually setting the UI label, but now generate that dynamically. This will cause old labels from 2.2.4 pipelines to be destroyed.

I also removed links, because they are too hard to simulate in cypress (and this is hopefully something canvas already tests for)

Developer's Certificate of Origin 1.1

       By making a contribution to this project, I certify that:

       (a) The contribution was created in whole or in part by me and I
           have the right to submit it under the Apache License 2.0; or

       (b) The contribution is based upon previous work that, to the best
           of my knowledge, is covered under an appropriate open source
           license and I have the right under that license to submit that
           work with modifications, whether created in whole or in part
           by me, under the same open source license (unless I am
           permitted to submit under a different license), as indicated
           in the file; or

       (c) The contribution was provided directly to me by some other
           person who certified (a), (b) or (c) and I have not modified
           it.

       (d) I understand and agree that this project and the contribution
           are public and that a record of the contribution (including all
           personal information I submit with it, including my sign-off) is
           maintained indefinitely and may be redistributed consistent with
           this project or the open source license(s) involved.

### CODE REVIEW

**Review by lresende — CHANGES_REQUESTED** [review]
- Please add the `Developer's Certificate of Origin 1.1` to your pr
- We should have the scenario described in the issue #1784 :
   - Have at least three nodes
   - Uses environment variables
   - Include local file that is used by the node impl
   - Produce a file that is used on a subsequent node impl
   - Uses Notebook, Python Script, R Script nodes
   - Use node content from another sibling folder

**Comment by lresende** [comment]
Could we actually have the tests in `tests/integration` and maybe something like `pipeline-format.ts` or `pipeline-backward-compatibility`.

**Comment by lresende** [comment]
Also, I am not sure what the snapshot mechanism does, from `The first run of this in a test will generate a snapshot of the pipeline file if it does not exist. If it exists, it will verify that the file matches the stored snapshot.` it seems to imply that the snapshot can be created dynamically, but what I really think we want is to have a set of pipeline examples `certified` to be valid (e.g. test/resources/valid_empty_pipeline.json, test/resources/valid_3_node_pipeline) and make sure that creating a new pipeline via the UI from the build env generates an equivalent pipeline, and by equivalent, I mean it's exactly the same except for unique fields like node coordinates, ids, etc.

**Comment by lresende** [comment]
From the changes highlighted:
- The first two I believe are only additions
- The third one might be a bug? 

```
Node properties no longer include unset properties in pipeline json:

-  "include_subdirectories": false,
-  "runtime_image": "",
```

Are these fields that have a default value and weren't changed by the user? 
runtime_image should be there otherwise pipeline can't be saved, no?

**Comment by bourdakos1** [comment]
@lresende 

> Could we actually have the tests in tests/integration and maybe something like pipeline-format.ts or pipeline-backward-compatibility.

The test cases are currently in `tests/integration/pipeline.ts`, but I can move them into a new file with one of the suggested file names, do you have a preference?

> Also, I am not sure what the snapshot mechanism does

I normally see snapshot testing in UI unit tests, this is the description of snapshots from jest: "Snapshot tests are a very useful tool whenever you want to make sure your UI does not change unexpectedly."

I think we can reuse this concept for the pipeline json, because I think we basically want to make sure it doesn't change unexpectedly, right?

The main idea is that we would write an integration test that creates a pipeline, and then call `cy.readFile('example.pipeline').matchesSnapshot()` as the assertion. The first time you run the test, it would generate the actual snapshot, this is the "certified" example and thats what you see in the `snapshots/example.snap`. This snapshot is never supposed to change and if it does, it is a test failure.

> The third one might be a bug?

Yea, I wasn't sure if this should be considered a bug or not. The field `include_subdirectories` has a default value of false, so it is omitted unless manually edited. `runtime_image` is unset with no default so it really should be null instead of an empty string I think. Omitting these values is still backwards compatible because 2.2.4 will automatically fill  `include_subdirectories` and  `runtime_image` with false and an empty string respectively.

**Comment by bourdakos1** [comment]
notes to self

```
elyra/elyra-example-anaconda-py3:latest
```

**Comment by bourdakos1** [comment]
@lresende this should be good to go once https://github.com/elyra-ai/pipeline-editor/pull/136 is released and we replace the codesandbox version

**Comment by bourdakos1** [comment]
@marthacryan @ajbozarth can you take a closer look at my file request logic to make sure I'm not missing anything

File: Makefile [hunk 1]
```diff
@@ -100,6 +100,13 @@ dev-link:
 	cd node_modules/@elyra/pipeline-editor && jupyter labextension link --no-build .
 	cd node_modules/@elyra/pipeline-services && jupyter labextension link --no-build .
 
+dev-unlink:
+	yarn unlink @elyra/pipeline-services
+	yarn unlink @elyra/pipeline-editor
+	jupyter labextension uninstall @elyra/pipeline-services
+	jupyter labextension uninstall @elyra/pipeline-editor
```
Comment (ajbozarth) [inline Makefile:107]: this is handled in `uninstall`, is there a reason we need another command?

File: packages/pipeline-editor/package.json [hunk 2]
```diff
@@ -37,8 +37,8 @@
   },
   "dependencies": {
     "@elyra/metadata-common": "^2.3.0-dev",
-    "@elyra/pipeline-editor": "^0.7.0",
-    "@elyra/pipeline-services": "^0.7.0",
+    "@elyra/pipeline-editor": "https://pkg.csb.dev/elyra-ai/pipeline-editor/commit/42fac9a8/@elyra/pipeline-editor",
+    "@elyra/pipeline-services": "https://pkg.csb.dev/elyra-ai/pipeline-editor/commit/42fac9a8/@elyra/pipeline-services",
```
Comment (ajbozarth) [inline packages/pipeline-editor/package.json:41]: is there a reason you're using the snapshots instead of the dev build? (just curious)

**Review by ajbozarth — COMMENTED** [review]
I'm still trying to wrap my head around what the changes to `onFileRequested` do and why, but figured I should comment the note's I have while I do

File: packages/pipeline-editor/package.json (hunk 2 above)
Comment (bourdakos1) [reply packages/pipeline-editor/package.json:41]: I was using the dev build when testing locally, but I just wanted to make sure everything still worked in ci with the bundled version

File: Makefile (hunk 1 above)
Comment (bourdakos1) [reply Makefile:107]: I think it's convenient to have a command to just unlink, but if others don't find it useful I can remove

Comment (marthacryan) [reply Makefile:107]: I think that'd be useful

Comment (ajbozarth) [reply Makefile:107]: fair enough I guess make sure to add `dev-unlink` to `.PHONY` since it's a new command

Comment (bourdakos1) [reply Makefile:107]: oh yea I think we also need to add dev-link to .PHONY too

Comment (ajbozarth) [reply Makefile:107]: thx, I had added it, but I must've been lost in some merge or rebase along the way

**Comment by lresende** [comment]
@bourdakos1 Where is this comparing the built pipeline with the `consumer-producer.pipeline` that was built on the previous Elyra release?

**Comment by lresende** [comment]
The complex.snapshot in this pr is not compatible with Elyra 2.2.4. If you compare the producer-consumer.pipeline and the snapshot you will see that the issue reported in #1693 is still there and the comparison is not picking it up.

**Comment by lresende** [comment]
You can launch Elyra 2.2.4 via docker with:
```
docker run -it -p 8888:8888 -v ${HOME}/opensource/jupyter-notebooks:/home/jovyan/work -v ${HOME}/Library/Jupyter:/home/jovyan/.local/share/jupyter -w /home/jovyan/work elyra/elyra:2.2.4 jupyter lab --debug
```

**Comment by bourdakos1** [comment]
@lresende it did pick up that change as noted above, but I didn’t notice the bug. I think this issue was fixed at one point, but must have been re-introduced when moving the component definitions back to the server. I think the fix for this is changing the cpu/gpu type to number instead of string. 

I can roll back the snapshot so the tests fail until fixed

**Review by lresende — APPROVED** [review]
Thanks @bourdakos1 , LGTM now.

**Pull request merged**